File: backend/app/services/custom_sql.py
"""Guardrails for custom-SQL checks (ADR 0019)."""

from __future__ import annotations

import re
import string
from typing import Any

from backend.app.core.errors import DataQError

# Trailing characters a single statement may end with (whitespace + a closing semicolon).
_TRAILING_CHARS = string.whitespace + ";"

# The GX expectation a custom-SQL check maps to (ADR 0019).
CUSTOM_SQL_EXPECTATION_TYPE = "unexpected_rows_expectation"
# The GX kwarg holding the user's query.
QUERY_KEY = "unexpected_rows_query"

# Datasources GX can run SQL against (ADR 0019). Flat files (adls_gen2 / s3) are
# DataFrame assets; orchestration types (adf / airflow) aren't datasources at all.
SQL_QUERYABLE_TYPES = frozenset({"snowflake", "unity_catalog"})

# Statement keywords that mutate data, schema, permissions, or transaction state.
_FORBIDDEN_KEYWORDS = frozenset(
    {
        "insert",
        "update",
        "delete",
        "merge",
        "upsert",
        "truncate",
        "drop",
        "alter",
        "create",
        "grant",
        "revoke",
        "commit",
        "rollback",
        "into",  # SELECT ... INTO <table> creates a table in some dialects
        # Defence-in-depth: other state-changing / proc-invoking statements.
        "call",
        "exec",
        "execute",
        "do",
        "copy",
        "lock",
        "set",
        "reset",
        "discard",
        "prepare",
        "deallocate",
        "vacuum",
        "analyze",
        "use",
        "attach",
        "detach",
        "unload",
    }
)

_LEADING_KEYWORD = re.compile(r"[\s(]*([a-zA-Z]+)")
_WORD = re.compile(r"[a-zA-Z_]+")


class CustomSqlInvalidError(DataQError):
    status_code = 422
    code = "custom_sql_invalid"


def is_custom_sql(expectation_type: str) -> bool:
    """True if `expectation_type` is the custom-SQL expectation (ADR 0019)."""
    return expectation_type == CUSTOM_SQL_EXPECTATION_TYPE


def _strip_noncode(sql: str) -> tuple[str, bool]:
    """Replace comments and string literals with spaces in a single left-to-right
    pass, leaving only executable code. Returns ``(code, well_formed)``.
    """
    out: list[str] = []
    i, n = 0, len(sql)
    well_formed = True
    while i < n:
        pair = sql[i : i + 2]
        if pair == "--":
            nl = sql.find("\n", i)
            i = n if nl == -1 else nl
            out.append(" ")
        elif pair == "/*":
            end = sql.find("*/", i + 2)
            if end == -1:
                well_formed = False  # unterminated block comment
                i = n
            else:
                i = end + 2
            out.append(" ")
        elif sql[i] in "'\"":
            quote = sql[i]
            i += 1
            closed = False
            while i < n:
                if sql[i] == quote:
                    if sql[i + 1 : i + 2] == quote:
                        i += 2  # doubled quote = escaped, stay in the string
                        continue
                    i += 1
                    closed = True
                    break
                i += 1
            if not closed:
                well_formed = False  # unterminated string literal
            out.append(" ")
        else:
            out.append(sql[i])
            i += 1
    return "".join(out), well_formed


def validate_query(raw_query: Any) -> None:
    """Reject a non-read-only or multi-statement custom-SQL query (422)."""
    if not isinstance(raw_query, str) or not raw_query.strip():
        raise CustomSqlInvalidError(
            f"custom-SQL check requires a non-empty {QUERY_KEY!r}",
            detail={"query_key": QUERY_KEY},
        )

    code, well_formed = _strip_noncode(raw_query)
    if not well_formed:
        # An unterminated string/comment means the rest of the query was swallowed
        # as literal text — we can't reason about it, so fail closed.
        raise CustomSqlInvalidError(
            "custom-SQL has an unterminated string literal or comment",
            detail={"query_key": QUERY_KEY},
        )

    analysis = code.strip().rstrip(_TRAILING_CHARS)
    if not analysis:
        raise CustomSqlInvalidError(
            "custom-SQL query is empty after removing comments",
            detail={"query_key": QUERY_KEY},
        )

    if ";" in analysis:
        raise CustomSqlInvalidError(
            "custom-SQL must be a single statement (no ';'-chained statements)",
            detail={"query_key": QUERY_KEY},
        )

    leading = _LEADING_KEYWORD.match(analysis)
    first_kw = leading.group(1).lower() if leading else ""
    if first_kw not in {"select", "with"}:
        raise CustomSqlInvalidError(
            "custom-SQL must be a read-only SELECT / WITH query",
            detail={"query_key": QUERY_KEY, "first_keyword": first_kw or None},
        )

    forbidden = sorted({w.lower() for w in _WORD.findall(analysis)} & _FORBIDDEN_KEYWORDS)
    if forbidden:
        raise CustomSqlInvalidError(
            "custom-SQL must be read-only; remove the disallowed keyword(s)",
            detail={"query_key": QUERY_KEY, "forbidden": forbidden},
        )


def validate_custom_sql_check(
    *, expectation_type: str, config: dict[str, Any], connection_type: str
) -> None:
    """Guardrail for a custom-SQL check; a no-op for any other expectation."""
    if not is_custom_sql(expectation_type):
        return
    if connection_type not in SQL_QUERYABLE_TYPES:
        raise CustomSqlInvalidError(
            f"custom-SQL checks require a SQL datasource, not {connection_type!r}",
            detail={
                "connection_type": connection_type,
                "supported": sorted(SQL_QUERYABLE_TYPES),
            },
        )
    validate_query(config.get(QUERY_KEY))

File: frontend/src/components/checks/customSql.ts
/**
 * Client-side mirror of the backend custom-SQL guardrail (ADR 0019) for inline
 * editor feedback. The backend `custom_sql.validate_query` is the authoritative
 * boundary (and runs with extra="forbid" + a least-privilege role); this is a
 * lighter, best-effort pre-check so the editor flags an obviously non-read-only
 * or multi-statement query before the save round-trips to a 422.
 *
 * Kept pure (no React) so it's unit- and mutation-testable in isolation.
 */

/** The GX expectation a custom-SQL check maps to (ADR 0019). */
export const CUSTOM_SQL_EXPECTATION_TYPE = 'unexpected_rows_expectation';
/** The GX kwarg / config key holding the user's query. */
export const CUSTOM_SQL_QUERY_KEY = 'unexpected_rows_query';

export const isCustomSql = (expectationType: string | undefined): boolean =>
  expectationType === CUSTOM_SQL_EXPECTATION_TYPE;

// State-mutating / proc-invoking statement keywords — a read-only check query
// must contain none of them (as a bareword, after strings/comments are stripped).
// `comment` / `replace` are omitted (common column name / `replace()` function),
// matching the backend set.
const FORBIDDEN_KEYWORDS = new Set([
  'insert',
  'update',
  'delete',
  'merge',
  'upsert',
  'truncate',
  'drop',
  'alter',
  'create',
  'grant',
  'revoke',
  'commit',
  'rollback',
  'into',
  'call',
  'exec',
  'execute',
  'do',
  'copy',
  'lock',
  'set',
  'reset',
  'discard',
  'prepare',
  'deallocate',
  'vacuum',
  'analyze',
  'use',
  'attach',
  'detach',
  'unload',
]);

/**
 * Replace comments and string literals with spaces in a single left-to-right pass
 * so neither can mask the other. Backtick is intentionally NOT a quote (Snowflake
 * / Unity Catalog don't delimit strings with it). Returns `null` if a string or
 * block comment is left unterminated — the caller fails closed.
 */
function stripNonCode(sql: string): string | null {
  let out = '';
  let i = 0;
  const n = sql.length;
  let wellFormed = true;
  while (i < n) {
    const pair = sql.slice(i, i + 2);
    if (pair === '--') {
      const nl = sql.indexOf('\n', i);
      i = nl === -1 ? n : nl;
      out += ' ';
    } else if (pair === '/*') {
      const end = sql.indexOf('*/', i + 2);
      if (end === -1) {
        wellFormed = false;
        i = n;
      } else {
        i = end + 2;
      }
      out += ' ';
    } else if (sql[i] === "'" || sql[i] === '"') {
      const quote = sql[i];
      i += 1;
      let closed = false;
      while (i < n) {
        if (sql[i] === quote) {
          if (sql[i + 1] === quote) {
            i += 2; // doubled quote = escaped, stay in the string
            continue;
          }
          i += 1;
          closed = true;
          break;
        }
        i += 1;
      }
      if (!closed) wellFormed = false;
      out += ' ';
    } else {
      out += sql[i];
      i += 1;
    }
  }
  return wellFormed ? out : null;
}

const TRAILING = /[\s;]+$/;
const LEADING_KEYWORD = /^[\s(]*([a-zA-Z]+)/;
const WORD = /[a-zA-Z_]+/g;

/**
 * Best-effort read-only / single-statement check. Returns an error message, or
 * `null` if the query looks acceptable. The backend is authoritative.
 */
export function validateCustomSqlQuery(query: string | undefined): string | null {
  if (!query || !query.trim()) return 'Enter a SQL query.';

  const code = stripNonCode(query);
  if (code === null) return 'Unterminated string literal or comment.';

  const analysis = code.trim().replace(TRAILING, '');
  if (!analysis) return 'Query is empty after removing comments.';
  if (analysis.includes(';')) return 'Use a single statement (no “;”-chained statements).';

  const first = LEADING_KEYWORD.exec(analysis)?.[1]?.toLowerCase();
  if (first !== 'select' && first !== 'with') {
    return 'Must be a read-only SELECT / WITH query.';
  }

  const words = analysis.toLowerCase().match(WORD) ?? [];
  const forbidden = words.find((w) => FORBIDDEN_KEYWORDS.has(w));
  if (forbidden) return `Must be read-only; remove “${forbidden}”.`;

  return null;
}
